File: static/js/charts/scatter.js
/**
 * D3 Scatter plot implementation for quadrant visualization.
 * Used in Essay 2: Not All Champions Are Safe
 */

function renderScatter(container, spec) {
    const { data, config } = spec;
    const { points } = data;

    const width = config.width || 700;
    const height = config.height || 500;
    const margin = { top: 40, right: 40, bottom: 60, left: 70 };
    const innerWidth = width - margin.left - margin.right;
    const innerHeight = height - margin.top - margin.bottom;

    d3.select(container).selectAll('*').remove();

    const svg = d3.select(container)
        .append('svg')
        .attr('width', '100%')
        .attr('height', height)
        .attr('viewBox', `0 0 ${width} ${height}`);

    const g = svg.append('g')
        .attr('transform', `translate(${margin.left},${margin.top})`);

    // Scales
    const xScale = d3.scaleLinear()
        .domain(config.xDomain || [0, d3.max(points, d => d.x) * 1.1])
        .range([0, innerWidth]);

    const yScale = d3.scaleLinear()
        .domain(config.yDomain || [0, 1])
        .range([innerHeight, 0]);

    // Quadrant backgrounds
    if (config.quadrants) {
        const xThresh = xScale(config.quadrants.x_threshold);
        const yThresh = yScale(config.quadrants.y_threshold);
        const colors = config.quadrants.quadrant_colors;

        // Top-left: Low value, high risk
        g.append('rect')
            .attr('x', 0).attr('y', 0)
            .attr('width', xThresh).attr('height', yThresh)
            .attr('fill', colors.top_left);

        // Top-right: High value, high risk (danger)
        g.append('rect')
            .attr('x', xThresh).attr('y', 0)
            .attr('width', innerWidth - xThresh).attr('height', yThresh)
            .attr('fill', colors.top_right);

        // Bottom-left: Low value, stable
        g.append('rect')
            .attr('x', 0).attr('y', yThresh)
            .attr('width', xThresh).attr('height', innerHeight - yThresh)
            .attr('fill', colors.bottom_left);

        // Bottom-right: High value, stable (ideal)
        g.append('rect')
            .attr('x', xThresh).attr('y', yThresh)
            .attr('width', innerWidth - xThresh).attr('height', innerHeight - yThresh)
            .attr('fill', colors.bottom_right);

        // Quadrant labels
        const labels = config.quadrants.quadrant_labels;
        g.append('text').attr('x', xThresh / 2).attr('y', 20).attr('text-anchor', 'middle')
            .attr('fill', '#718096').attr('font-size', '11px').text(labels.top_left);
        g.append('text').attr('x', xThresh + (innerWidth - xThresh) / 2).attr('y', 20)
            .attr('text-anchor', 'middle').attr('fill', '#e53e3e').attr('font-size', '11px')
            .attr('font-weight', 'bold').text(labels.top_right);
    }

    // Axes
    g.append('g')
        .attr('transform', `translate(0,${innerHeight})`)
        .call(d3.axisBottom(xScale).ticks(6).tickFormat(d => `$${d3.format(',.0f')(d)}`))
        .selectAll('text').attr('fill', '#4a5568');

    g.append('g')
        .call(d3.axisLeft(yScale).ticks(5).tickFormat(d3.format('.0%')))
        .selectAll('text').attr('fill', '#4a5568');

    // Axis labels
    g.append('text')
        .attr('x', innerWidth / 2).attr('y', innerHeight + 45)
        .attr('text-anchor', 'middle').attr('fill', '#4a5568')
        .text(config.xLabel || 'Revenue');

    g.append('text')
        .attr('transform', 'rotate(-90)')
        .attr('x', -innerHeight / 2).attr('y', -50)
        .attr('text-anchor', 'middle').attr('fill', '#4a5568')
        .text(config.yLabel || 'Fragility Score');

    // Points
    const dots = g.selectAll('.dot')
        .data(points)
        .join('circle')
        .attr('class', 'dot')
        .attr('cx', d => xScale(d.x))
        .attr('cy', d => yScale(d.y))
        .attr('r', config.pointRadius || 6)
        .attr('fill', d => d.color || (d.is_fragile ? '#fc8181' : '#68d391'))
        .attr('opacity', config.pointOpacity || 0.7)
        .attr('stroke', '#fff')
        .attr('stroke-width', 1)
        .on('mouseover', function(event, d) {
            d3.select(this).attr('r', (config.pointRadius || 6) * 1.5);
            showTooltip(event, `Revenue: $${d.x.toLocaleString()}<br>Fragility: ${(d.y * 100).toFixed(0)}%`);
        })
        .on('mouseout', function() {
            d3.select(this).attr('r', config.pointRadius || 6);
            hideTooltip();
        });

    return {
        highlightFragile: function() {
            dots.attr('opacity', d => d.is_fragile ? 1 : 0.2);
        },
        highlightQuadrant: function(quadrant) {
            // Highlight specific quadrant
            dots.attr('opacity', d => {
                const inHighRisk = d.y > (config.quadrants?.y_threshold || 0.5);
                const inHighValue = d.x > (config.quadrants?.x_threshold || 500);
                if (quadrant === 'top_right') return (inHighRisk && inHighValue) ? 1 : 0.2;
                return 0.7;
            });
        },
        showAll: function() {
            dots.attr('opacity', config.pointOpacity || 0.7);
        }
    };
}

if (typeof window !== 'undefined') {
    window.renderScatter = renderScatter;
}

File: src/essays/charts/scatter.py
"""
Scatter plot specification for quadrant visualizations.

Used in Essay 2: Not All Champions Are Safe
Shows revenue vs fragility for champion customers.
"""

from typing import Any

from src.essays.base import ChartSpec
from src.essays.data_queries import ChampionFragilityData


def create_scatter_spec(
    chart_id: str,
    data: ChampionFragilityData,
    *,
    width: int = 700,
    height: int = 500,
) -> ChartSpec:
    """Create a quadrant scatter plot specification.

    Args:
        chart_id: Unique identifier for the chart
        data: Champion fragility data with quadrant points
        width: Chart width in pixels
        height: Chart height in pixels

    Returns:
        ChartSpec for D3 scatter plot rendering
    """
    # Build point data
    points = []
    for point in data.quadrant_data:
        points.append({
            "x": point["x"],  # Revenue
            "y": point["y"],  # Fragility score
            "id": point["id"],
            "is_fragile": point["is_fragile"],
            "color": "#fc8181" if point["is_fragile"] else "#68d391",
        })

    # Calculate axis ranges
    revenues = [p["x"] for p in points] if points else [0]
    max_revenue = max(revenues) if revenues else 1000

    # Quadrant boundaries
    quadrant_config = {
        "x_threshold": max_revenue / 2,  # Mid-point for revenue
        "y_threshold": 0.5,  # Fragility threshold
        "quadrant_labels": {
            "top_left": "Low Value / High Risk",
            "top_right": "High Value / High Risk (Priority!)",
            "bottom_left": "Low Value / Stable",
            "bottom_right": "High Value / Stable (Ideal)",
        },
        "quadrant_colors": {
            "top_left": "rgba(252, 129, 129, 0.1)",
            "top_right": "rgba(252, 129, 129, 0.2)",  # Highlight danger zone
            "bottom_left": "rgba(160, 174, 192, 0.1)",
            "bottom_right": "rgba(104, 211, 145, 0.1)",
        },
    }

    # Scrolly triggers
    scrolly_triggers = [
        {
            "step": 1,
            "action": "show_points",
            "params": {"filter": "all", "description": "All champions"},
        },
        {
            "step": 2,
            "action": "show_quadrants",
            "params": {"description": "Revenue vs Fragility quadrants"},
        },
        {
            "step": 3,
            "action": "highlight_quadrant",
            "params": {
                "quadrant": "top_right",
                "description": "High-value, high-risk: these need immediate attention",
            },
        },
        {
            "step": 4,
            "action": "highlight_points",
            "params": {"filter": "fragile", "description": f"{data.fragile_champion_count} fragile champions"},
        },
    ]

    return ChartSpec(
        chart_id=chart_id,
        chart_type="scatter",
        data={"points": points},
        config={
            "width": width,
            "height": height,
            "xLabel": "Customer Revenue ($)",
            "yLabel": "Fragility Score",
            "xDomain": [0, max_revenue * 1.1],
            "yDomain": [0, 1],
            "pointRadius": 6,
            "pointOpacity": 0.7,
            "quadrants": quadrant_config,
            "tooltip": {
                "show": True,
                "fields": ["revenue", "fragility", "id"],
            },
        },
        scrolly_triggers=scrolly_triggers,
        annotations=[
            {
                "type": "horizontal_line",
                "y": 0.5,
                "label": "Risk threshold",
                "style": "dashed",
            },
        ],
    )


def create_thermometer_spec(
    chart_id: str,
    data: ChampionFragilityData,
    *,
    width: int = 300,
    height: int = 400,
) -> ChartSpec:
    """Create a risk thermometer visualization.

    Shows distribution of champions by risk level.

    Args:
        chart_id: Unique identifier
        data: Champion fragility data
        width: Chart width
        height: Chart height

    Returns:
        ChartSpec for thermometer visualization
    """
    # Risk buckets
    buckets = [
        {
            "level": "high",
            "label": "High Risk",
            "count": data.fragility_buckets.get("high", 0),
            "color": "#fc8181",
        },
        {
            "level": "medium",
            "label": "Medium Risk",
            "count": data.fragility_buckets.get("medium", 0),
            "color": "#f6ad55",
        },
        {
            "level": "low",
            "label": "Low Risk",
            "count": data.fragility_buckets.get("low", 0),
            "color": "#68d391",
        },
    ]

    total = sum(b["count"] for b in buckets)
    for bucket in buckets:
        bucket["percentage"] = (bucket["count"] / total * 100) if total > 0 else 0

    return ChartSpec(
        chart_id=chart_id,
        chart_type="thermometer",
        data={"buckets": buckets, "total": total},
        config={
            "width": width,
            "height": height,
            "orientation": "vertical",
            "showLabels": True,
            "showCounts": True,
            "animate": True,
        },
        scrolly_triggers=[
            {
                "step": 1,
                "action": "fill_thermometer",
                "params": {"duration": 1500},
            },
            {
                "step": 2,
                "action": "highlight_level",
                "params": {"level": "high", "pulse": True},
            },
        ],
    )
